### ISSUE TITLE
Phase 48.3: ExperienceReplayBuffer — Episodic Memory, Generative Replay & Gradient Constraints

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#940]
### Overview & Motivation

Replay-based methods combat catastrophic forgetting by maintaining and replaying data (or pseudo-data) from previous tasks during new task learning. The ExperienceReplayBuffer implements multiple replay strategies: **episodic memory replay** (storing raw exemplars), **generative replay** (training a generator to produce pseudo-exemplars), **gradient episodic memory** (constraining gradient directions), and **dark experience replay** (storing and replaying model logits).

These approaches are particularly effective for class-incremental learning (the hardest continual learning scenario) and can be combined with regularization methods for maximum performance.

### Architecture Components

```
ExperienceReplayBuffer
├── EpisodicMemory
│   ├── ReservoirSampler             # Reservoir sampling for balanced memory
│   ├── HerdingSelector              # iCaRL-style herding for exemplar selection
│   ├── KCenterCoresetSelector       # K-center coreset for coverage maximization
│   ├── EntropyBasedSelector         # High-uncertainty sample prioritization
│   └── ClassBalancedMemory          # Equal exemplars per class
├── GenerativeReplayEngine
│   ├── VAEReplayGenerator           # Variational autoencoder for pseudo-exemplars
│   ├── GANReplayGenerator           # GAN-based replay (Shin et al., 2017)
│   ├── ConditionalGenerator         # Class-conditional generation
│   ├── GeneratorUpdater             # Incrementally update generator with new data
│   └── QualityValidator             # Validate generated sample quality
├── GradientEpisodicMemory
│   ├── GEMProjector                 # Project gradients to avoid increasing episodic loss
│   ├── AGEMApproximator             # Averaged GEM (Chaudhry et al., 2019)
│   ├── ReferenceGradientComputer    # Compute reference gradients on memory
│   └── ViolationDetector            # Detect constraint violations
├── DarkExperienceReplay
│   ├── LogitStore                   # Store model logits alongside inputs
│   ├── DarkKnowledgeDistiller       # Knowledge distillation from stored logits
│   ├── AlphaScheduler               # Balance CE loss vs dark knowledge loss
│   └── LogitCompressor              # Compress stored logits for memory efficiency
├── BufferManager
│   ├── MemoryAllocator              # Per-task memory budget allocation
│   ├── EvictionPolicy               # Which samples to evict when buffer is full
│   ├── SamplingStrategy             # How to sample from buffer during training
│   ├── AugmentationPipeline         # Augment replayed samples for diversity
│   └── BufferStatistics             # Track buffer composition and usage
└── IntegrationLayer
    ├── ReplayScheduler              # When and how often to replay
    ├── BatchMixer                   # Mix current task data with replay data
    ├── LossBalancer                 # Balance current vs replay losses
    └── ReplayEffectivenessTracker   # Monitor replay impact on forgetting
```

### Key Algorithms & Data Structures

**Gradient Episodic Memory (GEM, Lopez-Paz & Ranzato, 2017):**
```python
class GEMProjector:
    def project_gradient(self, current_grad, reference_grads):
        """Project gradient to satisfy episodic memory constraints."""
        # g = current gradient, G = matrix of reference gradients
        # If g^T · g_ref < 0 for any ref, project:
        # min ||g_tilde - g||² s.t. g_tilde^T · g_ref >= 0 ∀ ref
        violations = [(i, ref) for i, ref in enumerate(reference_grads)
                      if torch.dot(current_grad, ref) < 0]
        if not violations:
            return current_grad  # No constraint violation
        # Solve QP: dual variables via quadprog
        G = torch.stack(reference_grads)
        dotprod = G @ current_grad
        GG = G @ G.T
        v = self.solve_qp(GG, dotprod)
        return current_grad - G.T @ v
```

**iCaRL Herding Selection (Rebuffi et al., 2017):**
```python
class HerdingSelector:
    def select_exemplars(self, features, budget_per_class):
        """Select exemplars closest to class mean in feature space."""
        class_mean = features.mean(dim=0)
        selected = []
        running_sum = torch.zeros_like(class_mean)
        for _ in range(budget_per_class):
            # Argmin ||class_mean - (running_sum + f) / (len(selected) + 1)||
            residual = class_mean - (running_sum + features) / (len(selected) + 1)
            idx = residual.norm(dim=1).argmin()
            selected.append(idx)
            running_sum += features[idx]
        return selected
```

**Reservoir Sampling (Vitter, 1985):**
```python
class ReservoirSampler:
    def add(self, sample, seen_count):
        if len(self.buffer) < self.capacity:
            self.buffer.append(sample)
        else:
            j = random.randint(0, seen_count)
            if j < self.capacity:
                self.buffer[j] = sample
```

**Dark Experience Replay (Buzzega et al., 2020):**
```python
class DarkExperienceReplay:
    def compute_loss(self, model, current_batch, replay_batch, alpha=0.5):
        # Standard CE on current data
        ce_loss = F.cross_entropy(model(current_batch.x), current_batch.y)
        # Dark knowledge on replay data (match stored logits)
        replay_logits = model(replay_batch.x)
        dark_loss = F.mse_loss(replay_logits, replay_batch.stored_logits)
        return ce_loss + alpha * dark_loss
```

### API Surface

```python
class ExperienceReplayBuffer:
    def __init__(self, config: ReplayConfig): ...
    def add_task_data(self, task_id: str, dataloader: DataLoader) -> None: ...
    def select_exemplars(self, method: str = "herding", budget: int = 200) -> ExemplarSet: ...
    def sample_replay_batch(self, batch_size: int) -> ReplayBatch: ...
    def compute_gem_gradient(self, current_grad: Tensor) -> Tensor: ...
    def store_logits(self, inputs: Tensor, logits: Tensor, task_id: str) -> None: ...
    def generate_pseudo_exemplars(self, task_id: str, num: int) -> Tensor: ...
    def get_buffer_stats(self) -> BufferStatistics: ...
    def mix_batch(self, current: Batch, replay_ratio: float = 0.5) -> MixedBatch: ...
    def evict(self, strategy: str = "reservoir") -> EvictionResult: ...
```

### Testing Strategy

- Reservoir sampling uniformity test (chi-squared test on sample distribution)
- Herding selection: verify selected exemplars approximate class mean
- GEM constraint satisfaction: project gradient must have non-negative dot product with all reference gradients
- Generative replay quality: FID score on generated vs real exemplars
- Dark experience replay: verify logit matching reduces KL divergence
- Buffer capacity enforcement: never exceed memory budget
- End-to-end: replay methods vs no-replay on Split-CIFAR-10

### References

1. Lopez-Paz & Ranzato (2017) — *Gradient Episodic Memory for Continual Learning*
2. Rebuffi et al. (2017) — *iCaRL: Incremental Classifier and Representation Learning*
3. Shin et al. (2017) — *Continual Learning with Deep Generative Replay*
4. Buzzega et al. (2020) — *Dark Experience for General Continual Learning*
5. Chaudhry et al. (2019) — *Efficient Lifelong Learning with A-GEM*
6. Vitter (1985) — *Random Sampling with a Reservoir*
7. Riemer et al. (2019) — *Learning to Learn without Forgetting by Maximizing Transfer and Minimizing Interference* (MER)

Part of **Phase 48: Continual Learning & Catastrophic Forgetting Prevention**
Discussion: #937
